### PR TITLE
support for @hidden directive

### PR DESCRIPTION
**Pull request by temaEmelyan**
Please make sure you consider the following:

- [ ] Add tests that use __typename in queries
- [ ] Does this change work with all nadel transformations (rename, type rename, hydration, etc)? Add tests for this.
- [ ] Is it worth using hints for this change in order to be able to enable a percentage rollout?
- [ ] Do we need to add integration tests for this change in the graphql gateway?
- [ ] Do we need a pollinator check for this?

### CODE REVIEW

File: api/src/main/java/graphql/nadel/Nadel.java [hunk 1]
```diff
@@ -69,7 +70,8 @@
     final Map<String, StitchingDsl> stitchingDsls;
     final ServiceExecutionFactory serviceExecutionFactory;
     final List<Service> services;
-    final GraphQLSchema overallSchema;
+    final GraphQLSchema privateOverallSchema;
```
Comment (bbakerman) [inline api/src/main/java/graphql/nadel/Nadel.java:73]: naming matters - its not private - LTE NOT NOT use this

Its meant to be the schema for "querying" and "introspection"

Comment (temaEmelyan) [inline api/src/main/java/graphql/nadel/Nadel.java:73]: I think we can extend  `NadelPublicGraphQLSchema` and `NadelPrivateGraphQLSchema` for better type safety (they can either extend GraphQLSchema or wrap it)

File: api/src/main/java/graphql/nadel/schema/PublicSchemaGenerator.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import graphql.schema.GraphQLDirective;
+import graphql.schema.GraphQLDirectiveContainer;
+import graphql.schema.GraphQLSchema;
+import graphql.schema.transform.FieldVisibilitySchemaTransformation;
+import graphql.schema.transform.VisibleFieldPredicate;
+
+import static graphql.nadel.schema.NadelDirectives.HIDDEN_DIRECTIVE_DEFINITION;
+
+@Internal
+public class PublicSchemaGenerator {
```
Comment (bbakerman) [inline api/src/main/java/graphql/nadel/schema/PublicSchemaGenerator.java:13]: Not public - not private

QuerySchema and OverallSchema

File: engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/blueprint/NadelExecutionBlueprint.kt [hunk 3]
```diff
@@ -14,7 +14,8 @@ import graphql.schema.GraphQLSchema
  * Execution blueprint where keys are in terms of the overall schema.
  */
 data class NadelOverallExecutionBlueprint(
-    val schema: GraphQLSchema,
+    val privateSchema: GraphQLSchema,
```
Comment (bbakerman) [inline engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/blueprint/NadelExecutionBlueprint.kt:17]: querySchema and overallSchema

File: api/src/main/java/graphql/nadel/schema/NadelDirectives.java [hunk 4]
```diff
@@ -273,7 +274,11 @@
                                 .build())
                 .build();
 
-
+        HIDDEN_DIRECTIVE_DEFINITION = DirectiveDefinition.newDirectiveDefinition()
+                .name("hidden")
+                .description(createDescription("Indicates that the field is not available for queries"))
```
Comment (bbakerman) [inline api/src/main/java/graphql/nadel/schema/NadelDirectives.java:279]: Indicates that the field is not available for queries or introspection

File: api/src/main/java/graphql/nadel/schema/PublicSchemaGenerator.java (hunk 2 above)
Comment (bbakerman) [reply api/src/main/java/graphql/nadel/schema/PublicSchemaGenerator.java:13]: naming discussed online

File: test/src/test/resources/fixtures/hidden/can-query-non-hidden-fields.yml [hunk 5]
```diff
@@ -0,0 +1,71 @@
+name: can query non hidden fields
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hidden/can-query-non-hidden-fields.yml:1]: Unrelated to this diff, but can we add another test for `@hidden` fields not at the top level?

Only thing I'd be worried about testing there is that the top level field routing is somehow being the one handling the hidden fields, then nested fields are never checked.

e.g. we should have a test like

```
type Query {
  hello: World
}

type World {
  area51: Area @hidden
}
```

where the query is

```
{
  hello {
    area51
  }
}
```

File: test/src/test/resources/fixtures/hidden/introspection-does-not-show-hidden-fields.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
+name: introspection does not show hidden fields
+enabled:
+  current: false
+  nextgen: true
+overallSchema:
+  MyService: |
+    directive @hidden on FIELD_DEFINITION
+    type Query {
+      hello: World
+      hiddenField: String @hidden
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hidden/introspection-does-not-show-hidden-fields.yml:10]: Nit: can we sprinkle a few more hidden fields on some other types?

File: api/src/main/java/graphql/nadel/Nadel.java [hunk 7]
```diff
@@ -215,17 +218,17 @@ public GraphQLSchema getOverallSchema() {
 
         NadelExecutionParams nadelExecutionParams = new NadelExecutionParams(nadelExecutionInput.getArtificialFieldsUUID(), nadelExecutionInput.getNadelExecutionHints());
 
-        InstrumentationState instrumentationState = instrumentation.createState(new NadelInstrumentationCreateStateParameters(overallSchema, executionInput));
-        NadelInstrumentationQueryExecutionParameters instrumentationParameters = new NadelInstrumentationQueryExecutionParameters(executionInput, overallSchema, instrumentationState);
+        InstrumentationState instrumentationState = instrumentation.createState(new NadelInstrumentationCreateStateParameters(querySchema, executionInput));
+        NadelInstrumentationQueryExecutionParameters instrumentationParameters = new NadelInstrumentationQueryExecutionParameters(executionInput, querySchema, instrumentationState);
```
Comment (gnawf) [inline api/src/main/java/graphql/nadel/Nadel.java:222]: Hmmm, wondering whether it's correct to pass in engine or query schema here. I'm leaning towards engine schema, but I see that this is also before any validation has occurred… hmmm.

File: engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/query/NadelFieldToService.kt [hunk 8]
```diff
@@ -91,7 +92,7 @@ internal class NadelFieldToService(
     }
 
     private fun isNamespacedField(field: ExecutableNormalizedField): Boolean {
-        return isNamespacedField(field, overallExecutionBlueprint.schema)
+        return isNamespacedField(field, querySchema)
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/query/NadelFieldToService.kt:95]: I think we should swap this back to engine schema. If we stick to the rule that

1. Query schema for anything user facing (validation, introspection).
2. Engine schema for almost everything else.

It's a lot easier to understand what to use. We should add a comment about that on `querySchema` somewhere too.

**Review by gnawf — CHANGES_REQUESTED** [review]
Some comments about usage of `engineSchema` vs `querySchema`. Also a couple more tests would be great.

Happy to approve after :)

**Review by gnawf — APPROVED** [review]
Nice

**Pull request merged**